src/examples/mnist/MNIST_SP.cpp

```diff
@@ -0,0 +1,240 @@
+#include <vector>
+
+#include <nupic/algorithms/SpatialPoolerExtended.hpp>
+#include <nupic/algorithms/SDRClassifier.hpp>
+#include <nupic/algorithms/ClassifierResult.hpp>
+#include <nupic/utils/Random.hpp>
+#include <nupic/ntypes/SdrMetrics.hpp>
+
+using namespace std;
+using namespace nupic;
+using nupic::algorithms::spatial_pooler_extended::SpatialPoolerExtended;
+using nupic::algorithms::sdr_classifier::SDRClassifier;
+using nupic::algorithms::cla_classifier::ClassifierResult;
+
+
+vector<UInt> read_mnist_labels(string path) {
+    ifstream file(path);
+    if( !file.is_open() ) {
+        cerr << "ERROR: Failed to open file " << path << endl;
+        exit(1);
+    }
+    int magic_number     = 0;
+    int number_of_labels = 0;
+    file.read( (char*) &magic_number,     4);
+    file.read( (char*) &number_of_labels, 4);
+    if(magic_number != 0x00000801) {
+        std::reverse((char*) &magic_number,      (char*) &magic_number + 4);
+        std::reverse((char*) &number_of_labels,  (char*) &number_of_labels + 4);
+    }
+    if(magic_number != 0x00000801) {
+        cerr << "ERROR: MNIST data is compressed or corrupt" << endl;
+        exit(1);
+    }
+    vector<UInt> retval;
+    for(int i = 0; i < number_of_labels; ++i) {
+        unsigned char label = 0;
+        file.read( (char*) &label, 1);
+        retval.push_back((UInt) label);
+    }
+    return retval;
+}
+
+
+vector<UInt*> read_mnist_images(string path) {
+    ifstream file(path);
+    if( !file.is_open() ) {
+        cerr << "ERROR: Failed to open file " << path << endl;
+        exit(1);
+    }
+    int magic_number     = 0;
+    int number_of_images = 0;
+    int n_rows           = 0;
+    int n_cols           = 0;
+    file.read( (char*) &magic_number,     4);
+    file.read( (char*) &number_of_images, 4);
+    file.read( (char*) &n_rows,           4);
+    file.read( (char*) &n_cols,           4);
+    if(magic_number != 0x00000803) {
+        std::reverse((char*) &magic_number,      (char*) &magic_number + 4);
+        std::reverse((char*) &number_of_images,  (char*) &number_of_images + 4);
+        std::reverse((char*) &n_rows,            (char*) &n_rows + 4);
+        std::reverse((char*) &n_cols,            (char*) &n_cols + 4);
+    }
+    if(magic_number != 0x00000803) {
+        cerr << "ERROR: MNIST data is compressed or corrupt" << endl;
+        exit(1);
+    }
+    NTA_ASSERT(n_rows == 28);
+    NTA_ASSERT(n_cols == 28);
+    UInt img_size = n_rows * n_cols;
+    vector<UInt*> retval;
+    for(int i = 0; i < number_of_images; ++i) {
+        auto data_raw = new unsigned char[img_size];
+        file.read( (char*) data_raw, img_size);
+        // Copy the data into an array of UInt's
+        auto data = new UInt[img_size];
+        // auto data = new UInt[2 * img_size];
+        // Apply a threshold to the image, yielding a B & W image.
+        for(UInt pixel = 0; pixel < img_size; pixel++) {
+            data[pixel] = data_raw[pixel] >= 128 ? 1 : 0;
```

Comment:
OT: how does this perform w/o the b/w reduction? (keeping the 256 colors)

Reply:
The MNIST dataset is mostly black & white, so more than 2 bits per input pixel doesn't  help.  2 bits per pixel gets about the same score as 1, but it takes longer to compute.  

With 2 bits per pixel, the input has constant sparsity, which is neat but the SP can deal with variable sparsity just fine so it's not really relevant.